### PR TITLE
Add pretty-line-printer

### PR DESCRIPTION
**Pull request by thumbnail**
## Brief

image

Similar to pretty-printer, but prints one line per report for a more concise output

Analogous to github.com/xojs/xo.

## QA plan

1. postfix version project.clj with -dev
2. lein install
3. `(require '[formatting-stack.project-formatter] '[formatting-stack.reporters.pretty-line-printer])`
3. run `(formatting-stack.project-formatter/format-and-lint-project! :reporter (formatting-stack.reporters.pretty-line-printer/new {}))`

## Author checklist

* [ ] I have QAed the functionality
* [x] The PR has a reasonably reviewable size and a meaningful commit history
* [x] I have run the [branch formatter](https://github.com/nedap/formatting-stack/blob/332a419034ab46fad526a5592f4257353bd695b6/src/formatting_stack/branch_formatter.clj) and observed no new/significative warnings
* [ ] The build passes
* [x] I have self-reviewed the PR prior to assignment
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [x] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [x] Test coverage
  * [ ] Spec coverage
  * [x] Documentation
  * [ ] Security
  * [x] Performance
  * [x] Breaking API changes
  * [x] Cross-compatibility (Clojure/ClojureScript)

## Reviewer checklist

* [x] I have checked out this branch and reviewed it locally, running it
* [x] I have QAed the functionality
* [x] I have reviewed the PR
* Additionally, I have code-reviewed iteratively the PR considering the following aspects in isolation:
  * [ ] Correctness
  * [ ] Robustness (red paths, failure handling etc)
  * [ ] Test coverage
  * [ ] Spec coverage
  * [ ] Documentation
  * [ ] Security
  * [ ] Performance
  * [ ] Breaking API changes
  * [ ] Cross-compatibility (Clojure/ClojureScript)

### CODE REVIEW

**Review by vemv — APPROVED** [review]
LGTM as it simply adds an opt-in feature

Perhaps I'd add some tests. A good edge case is what happens when things "overflow" (filename + reported fault exceed `max-msg-length`)

**Review by tcoenraad — APPROVED** [review]
Succesfully QA'd, except for missing emojis. 🤔 `(println "👍")` does work btw, so I'd say my terminal is emoji friendly.

image

**Review by tcoenraad — APPROVED** [review]
Re-QA'd, now the emojiis show up as expected

File: test/unit/formatting_stack/reporters/pretty_line_printer.clj [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                               (is (= expected
+                                      (with-out-str
+                                        (sut/print-warnings {:colorize? false
+                                                             :max-msg-length 20}
+                                                            input))))
+                               true)
+    "Empty"
+    []
+    ""
+
```
Comment (vemv) [inline test/unit/formatting_stack/reporters/pretty_line_printer.clj:17]: Looks like there could be a row exercising truncation

File: src/formatting_stack/reporters/pretty_line_printer.clj [hunk 2]
```diff
@@ -0,0 +1,78 @@
+(ns formatting-stack.reporters.pretty-line-printer
```
Comment (vemv) [inline src/formatting_stack/reporters/pretty_line_printer.clj:1]: Feel free to reference github.com/xojs/xo in the ns doc

...and similarly feel free to reference this ns from https://github.com/nedap/formatting-stack#consolidated-reporting or such (to keep these features discoverable)

**Review by vemv — APPROVED** [review]
LGTM!

File: src/formatting_stack/reporters/impl.clj [hunk 3]
```diff
@@ -17,3 +17,19 @@
                                        (string/replace (re-pattern (str "^" (System/getProperty "user.dir")))
                                                        "")
                                        (string/replace #"^/" ""))))
+
+(speced/defn truncate-line-wise [^string? s, length]
+  (if (= s "\n")
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/impl.clj:22]: Whats with this special case?

File: src/formatting_stack/reporters/impl.clj [hunk 4]
```diff
@@ -17,3 +17,19 @@
...
+         (map (fn [s]
+                (let [suffix "…"
+                      string-length (count s)
+                      suffix-length (count suffix)]
+                  (if (<= string-length length)
+                    s
+                    (str (subs s
+                               0
+                               (- length suffix-length))
+                         suffix)))))
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/impl.clj:34]: I guess you could `defn` this, overflow'ing a single line sounds useful on its own.

File: src/formatting_stack/reporters/pretty_line_printer.clj [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+   [nedap.speced.def :as speced]
+   [nedap.utils.modular.api :refer [implement]]))
+
+(speced/defn print-summary [{:keys [^boolean? summary?
+                                    ^boolean? colorize?]} reports]
+  (when summary?
+    (->> reports
+         (group-by :level)
+         (map-vals count)
+         (into (sorted-map-by compare)) ;; print summary in order
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/pretty_line_printer.clj:19]: TIL there is a treemap impl in clojure.core.

File: src/formatting_stack/reporters/pretty_line_printer.clj [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                                  :warning   " warnings found"))
+                   colorize? (colorize (case report-type
+                                         :exception :red
+                                         :error     :red
+                                         :warning   :yellow))
+                   true      println))))))
+
+(speced/defn print-exceptions [{:keys [^boolean? print-stacktraces?]} reports]
+  (->> reports
+       (filter (speced/fn [{:keys [^::protocols.spec/level level]}]
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/pretty_line_printer.clj:33]: Perhaps a bit late to start doing spec checks, would have personally defined a spec for `reports` I guess.

Also, lots of characters for `(comp #{:exception) :level)` ;).

File: src/formatting_stack/reporters/pretty_line_printer.clj [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+               (println (colorize title :cyan))
+               (doseq [{:keys [^Throwable exception]} reports]
+                 (if print-stacktraces?
+                   (print-stack-trace exception)
+                   (println (ex-message exception)))
+                 (println))))))
+
+(speced/defn print-warnings [{:keys [max-msg-length
+                                     ^boolean? colorize?]}
+                             ^::protocols.spec/reports reports]
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/pretty_line_printer.clj:46]: Hey look, the spec I was looking for!

File: src/formatting_stack/reporters/pretty_line_printer.clj [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  (->> (filter (fn [{:keys [level]}] (#{:error :warning} level)) reports)
+       (group-by :filename)
+       (into (sorted-map-by compare))
+       (run! (fn [[title reports]]
+               (println (cond-> title colorize? (colorize :cyan)))
+               (doseq [{:keys [msg msg-extra-data column line source level]} (sort-by :line reports)]
+                 (println (case level
+                            :error   " ❌"
+                            :warning " ⚠️")
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/pretty_line_printer.clj:55]: Is this GH being GH, or is one in an emoji, and the other .. not?

File: src/formatting_stack/reporters/pretty_line_printer.clj [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                          (if (or column line)
+                            (cond-> (format "%3s:%-3s" (or line "?") (or column "?")) colorize? (colorize :grey))
+                            "       ")
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/pretty_line_printer.clj:58]: Perhaps a bit overly brancing, would personally not mind `?:?` as a line:col number.

File: src/formatting_stack/reporters/pretty_line_printer.clj [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                          (format (str "%-" max-msg-length "." max-msg-length "s") msg)
+                          (cond-> (str "  " source) colorize? (colorize :grey)))
```
Comment (lennartbuit) [inline src/formatting_stack/reporters/pretty_line_printer.clj:60]: Maybe build the string, then conditionally colorize? Less branching like that.

**Review by lennartbuit — APPROVED** [review]
Some 🚲 🏚️

File: src/formatting_stack/reporters/pretty_line_printer.clj (hunk 6 above)
Comment (thumbnail) [reply src/formatting_stack/reporters/pretty_line_printer.clj:33]: > Perhaps a bit late to start doing spec checks, would have personally defined a spec for reports I guess.


There's. the report-protocol actually checks this already.

File: src/formatting_stack/reporters/pretty_line_printer.clj (hunk 8 above)
Comment (thumbnail) [reply src/formatting_stack/reporters/pretty_line_printer.clj:55]: There's a utf-8 character for ⚠️, which nowadays renders as a emoij. Pretty weird stuff. But it looks like an emoij most of the time.

File: src/formatting_stack/reporters/pretty_line_printer.clj (hunk 10 above)
Comment (thumbnail) [reply src/formatting_stack/reporters/pretty_line_printer.clj:60]: The code is mostly styled after `pretty-printer`; but i can take a look

File: src/formatting_stack/reporters/impl.clj (hunk 3 above)
Comment (thumbnail) [reply src/formatting_stack/reporters/impl.clj:22]: It's here to make sure` "\n"` isn't morphed into `""`:
```clj
user=> (clojure.string/join "\n" (clojure.string/split "\n" #"\n"))
""
```

Not sure what the origin is (or when `"/n"` would be passed into this fn to begin with). Would need to check.
